txp/Pos/Binary/Txp/Core.hs

```diff
@@ -19,10 +20,18 @@ import qualified Pos.Txp.Core.Types as T
 ----------------------------------------------------------------------------
 
 instance Bi T.TxIn where
-  encode txIn = encodeListLen 2 <> encode (T.txInHash txIn) <> encode (T.txInIndex txIn)
+  -- [CSL-1526] Support the "unknown" pattern for encoding a `TxIn`, so that
+  -- we can down the line extend the `TxIn` type is a backward-compatible way.
+  encode txIn =  encodeListLen 3
+              <> encode (0 :: Word8)
+              <> encode (T.txInHash txIn)
+              <> encode (T.txInIndex txIn)
   decode = do
-    enforceSize "TxIn" 2
-    T.TxIn <$> decode <*> decode
+    enforceSize "TxIn" 3
+    tag <- decode @Word8
+    case tag of
+        0 -> T.TxIn <$> decode <*> decode
```

Comment:
Wouldn't this break the symmetric nature of the encoding/decoding? When we encode, the first thing we do is to call `encodeListLen`, which writes in the serialised bytes the special token for "list is beginning" plus the size of the list.

Therefore, we must call `enforceSize` first thing during decoding, because that would (under the hood) try to read from the raw bytes the "list is beginning" token plus the size, and doing it at any other point of the decoding would make the process fail. In a nutshell, it won't roundtrip. Am I missing something?